false-positive return-in-except-star for lambdas

**Bug Report**

The `"return" not allowed in except* block  [misc]` check introduced in https://github.com/python/mypy/issues/18123 triggers when a lambda is defined inside an except* block. It does not trigger on nested functions defined in the same place.

**To Reproduce**

[mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=79489e17720a8f1f9c8b982474d043f0)

**Expected Behavior**

I expect mypy to not complain about the presence of return statements if a lambda is present in an except* block.

**Actual Behavior**

main.py:6: error: "return" not allowed in except* block  [misc]
main.py:9: error: "return" not allowed in except* block  [misc]
Found 2 errors in 1 file (checked 1 source file)

**Your Environment**

- Mypy version used: 1.15.0
- Python version used: 3.12
- Whatever mypy playground currently uses by default

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -6134,7 +6134,8 @@ class SemanticAnalyzer(
 
     def visit_lambda_expr(self, expr: LambdaExpr) -> None:
         self.analyze_arg_initializers(expr)
-        self.analyze_function_body(expr)
+        with self.inside_except_star_block_set(False, entering_loop=False):
+            self.analyze_function_body(expr)
 
     def visit_conditional_expr(self, expr: ConditionalExpr) -> None:
         expr.if_expr.accept(self)

--- a/test-data/unit/check-python311.test
+++ b/test-data/unit/check-python311.test
@@ -260,6 +260,24 @@ def foo():
             return  # E: "return" not allowed in except* block
 [builtins fixtures/exception.pyi]
 
+[case testLambdaInExceptStarBlock]
+# flags: --python-version 3.11
+def foo():
+    try:
+        pass
+    except* Exception:
+        x = lambda: 0
+        return lambda: 0  # E: "return" not allowed in except* block
+
+def loop():
+    while True:
+        try:
+            pass
+        except* Exception:
+            x = lambda: 0
+            return lambda: 0  # E: "return" not allowed in except* block
+[builtins fixtures/exception.pyi]
+
 [case testRedefineLocalWithinExceptStarTryClauses]
 # flags: --allow-redefinition
 def fn_str(_: str) -> int: ...

